release-19.2: changefeedccl: fix cloud file naming to prevent ordering violation across schema changes

danhhz opened the pull request:
Backport 1/1 commits from #41615.

/cc @cockroachdb/release

---

Currently, our naming convention for cloud storage sink files is
as follows:

```
<timestamp>-<uniquer>-<topic_id>-<schema_id>.<ext>
```

With `<uniquer>` being `<sessionid>-<nodeid>-<sinkid>-<fileid>`.

Now consider the following case where this convention fails to
preserve row orderings in presence of a schema change:
1. The sink starts buffering a file for schema `1`.
2. It then starts buffering a file for schema `2`.
3. The newer, schema `2` file exceeds the file size threshold
and thus gets flushed at timestamp `x`. This would lead to it being
assigned a `fileid` of `0`.
4. The older, schema `1` file is flushed at timestamp `x` and
assigned a `fileid` of `1`.

This leads to the older, schema `1` file, being lexically ordered
*after* the newer, schema `2` file.

This PR fixes this violation by making sure that when flushing a file
that corresponds to schema `x`, we flush all files in memory that
correspond to schema versions less than `x` before, in sorted order
of schema versions.

Release note: Fix bug in cloud storage sink for CDC.

Release justification: Fix bug in cloud storage sink file naming that
violates ordering in presence of a schema changes.

aayushshah15 committed:
changefeedccl: fix cloud storage file naming to preserve orderings
across schema changes

Currently, our naming convention for cloud storage sink files is
as follows:

```
<timestamp>-<uniquer>-<topic_id>-<schema_id>.<ext>
```

With `<uniquer>` being `<sessionid>-<nodeid>-<sinkid>-<fileid>`.

Now consider the following case where this convention fails to
preverse row orderings in presence of a schema change:
1. The sink starts buffering a file for schema `1`.
2. It then starts buffering a file for schema `2`.
3. The newer, schema `2` file exceeds the file size threshold
and thus gets flushed at timestamp `x`. This would lead to it being
assigned a `fileid` of `0`.
4. The older, schema `1` file is flushed at timestamp `x` and
assigned a `fileid` of `1`.

This leads to the older, schema `1` file, being lexically ordered
before the newer, schema `2` file.

This PR fixes this violation by making sure that when flushing a file
that corresponds to schema `x`, we flush all files in memory that
correspond to schema versions less than `x` before, in sorted order
of schema versions.

Release note (changefeedccl): Fix bug in cloud storage sink for CDC.

Release justification: Fix bug in cloud storage sink file naming that
violates ordering in presence of a schema changes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41693)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41693#-:-LrPkDZj-Mzi3PNRH8vz:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajwerner)



<!-- Sent from Reviewable.io -->

danhhz committed:
changefeedccl: fix backport merge skew

makeCloudStorageSink doesn't take an externalStorageFromURI arguement in
the release-19.2 branch.

Release note: None

danhhz commented:
TFTR!
